### PR TITLE
Fix #21722

### LINKED ISSUE #21722: returning string inside a loop unnecessarily generates `=sink`
### Description

This generate a simple assignment:
```nim
import std/strutils


func hey*(sentence: string): string =
  var i = sentence.high
  while true:
    if i < 0:
      break
    if sentence[i] notin Whitespace:
      break
    dec i

  if i < 0:
    return "Fine. Be that way!".
```
```nim
var i = high(sentence)
block :tmp:
  while true:
    if i < 0:
      break :tmp
    if not contains({' ', '\t', '\v', '\r', '\n', '\f'}, sentence[i]):
      break :tmp
    dec i, 1
if i < 0:
  return result = "Fine. Be that way!"
```
While this generates a `=sink`:
```nim
import std/strutils


func hey*(sentence: string): string =
  var i = sentence.high
  while true:
    if i < 0:
      return "Fine. Be that way!"
    if sentence[i] notin Whitespace:
      break
    dec i
```
```nim
var i = high(sentence)
block :tmp:
  while true:
    if i < 0:
      return `=sink`(result, "Fine. Be that way!")
    if not contains({' ', '\t', '\v', '\r', '\n', '\f'}, sentence[i]):
      break :tmp
    dec i, 1
```
See the generated [assembly](https://godbolt.org/z/x5s81WdYv).

### Nim Version

ed7c6cdc984b12a30a44344167083ec0dc46cf9a

### Current Output

_No response_

### Expected Output

_No response_

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by SirOlaf**
Close #21722

### CODE REVIEW

**Comment by Araq** [comment]
What if it's `result = f(); let x = g(); return (result = x #[needs =sink here]#; result)`

**Comment by SirOlaf** [comment]
That example works by chance because `handleNestedTempl` resets it to `normal`, but yeah it seems quite unreliable in its current state. Simplest solution would be adding a bool param to `p` to keep track of returns for a single layer, but not sure if that's better or worse

File: compiler/injectdestructors.nim [hunk 1]
```diff
@@ -79,11 +79,11 @@ proc getTemp(c: var Con; s: var Scope; typ: PType; info: TLineInfo): PNode =
 proc nestedScope(parent: var Scope; body: PNode): Scope =
   Scope(vars: @[], locals: @[], wasMoved: @[], final: @[], body: body, needsTry: false, parent: addr(parent))
 
-proc p(n: PNode; c: var Con; s: var Scope; mode: ProcessMode; tmpFlags = {sfSingleUsedTemp}): PNode
+proc p(n: PNode; c: var Con; s: var Scope; mode: ProcessMode; tmpFlags = {sfSingleUsedTemp}; inReturn = false): PNode
 
 type
   MoveOrCopyFlag = enum
-    IsDecl, IsExplicitSink
+    IsDecl, IsExplicitSink, IsReturn
```
Comment (Araq) [inline compiler/injectdestructors.nim:86]:
```suggestion
    IsDecl, IsExplicitSink
```

**Comment by Araq** [comment]
I screwed up! Please re-add the enum state.

**Pull request merged**